### PR TITLE
fix #24968 more friend error message for  `Self` in fn args

### LINKED ISSUE #24968: Wrong error message when using `Self` in some places
```rust
fn foo(_: Self) {}

fn main() {}
```
```
<anon>:1:11: 1:15 error: use of undeclared type name `Self`
<anon>:1 fn foo(_: Self) {}
                   ^~~~
```

`Self` is a keyword, this error message indicates that it can be used to define type, which is misleading.

### PR DESCRIPTION
**Pull request by XuefengWu**
fix https://github.com/rust-lang/rust/issues/24968
report more friendly error message for `Self` when fn args

### CODE REVIEW

**Comment by arielb1** [comment]
Your message sounds like a type error. I would prefer one in the style of "use of Self outside of an impl".

**Comment by nrc** [comment]
Hi @XuefengWu, thanks for the PR and sorry it took so long for me to get to it - I was on vacation last week.

A high level thing is that this PR needs a test to check that we get the right error messages. You should add a test to the compile-fail directory which checks the message.

More comments inline.

File: src/librustc_resolve/lib.rs [hunk 1]
```diff
@@ -2361,8 +2361,18 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
...
 
-                        let msg = format!("use of undeclared {} `{}`", kind,
-                                          path_names_to_string(path, 0));
+                        let self_type_name = special_idents::type_self.name;
+                        let is_invalid_self_type_name =
+                            path.segments.len() > 0 &&
+                            maybe_qself.is_none() &&
+                            path.segments[0].identifier.name == self_type_name;
+                        let msg = if is_invalid_self_type_name {
+                            "use of Self outside of an impl".to_string()
```
Comment (nrc) [inline src/librustc_resolve/lib.rs]: `Self` can also be used in a trait, so the message should include that. Also, you could add backticks around "Self".

**Comment by XuefengWu** [comment]
@nrc The change commit is: https://github.com/XuefengWu/rust/commit/eac661c165f31ff190d9abdec2e475f857d4a645

**Comment by nrc** [comment]
@XuefengWu something has gone wrong with the Git history here. I'm afraid we can't merge this as is. On local branches, it is usually better to use `rebase` rather than `merge` (`merge` can work too, but it is easier to get wrong and the history doesn't look as nice). To fix this, I would recommend starting a new branch from master and `cherry-pick`ing the commits from this branch. Then force push that branch to this remote: `git push origin new_branch_name:24968_err_msg_parse_self_type`. You can use `git rebase -i` to squash commits together or to reword commit messages (I note that the last commit here has a typo in 'keyword'). Let me know if you need more help with the git stuff, or ask on #rust-internals.

**Comment by nrc** [comment]
r=me with the git history sorted and the two nits fixed.

**Comment by nrc** [comment]
Closing in favour of #25485

**Comment by XuefengWu** [comment]
@nrc I copied a new PR https://github.com/rust-lang/rust/pull/25485

Should I close this PR?

**Pull request closed without merging**